src/util.cc

```diff
@@ -1,8 +1,187 @@
+  uint8_t acc;
+  const uint8_t* srcptr = input + length;
+  switch (length) {
+    default: return false;
+    case 4:
+      acc = (*--srcptr);
+      if (acc < 0x80 || acc > 0xBF)
+        return false;
+      // fall-through
+    case 3:
+      acc = (*--srcptr);
+      if (acc < 0x80 || acc > 0xBF)
+        return false;
+      // fall-through
+    case 2:
+      acc = (*--srcptr);
+      if (acc > 0xBF)
+        return false;
+      switch (*input) {
+        case 0xE0:
+          if (acc < 0xA0)
+            return false;
+          break;
+        case 0xED:
+          if (acc > 0x9F)
+            return false;
+          break;
+        case 0xF0:
+          if (acc < 0x90)
+            return false;
+          break;
+        case 0xF4:
+          if (acc > 0x8F)
+            return false;
+          break;
+        default:
+          if (acc < 0x80)
+            return false;
+      }
+      // fall-through
+    case 1:
+      if (*input >= 0x80 && *input < 0xC2) {
+        return false;
+      }
+  }
+  return *input <= 0xF4;
+}
+
+
+static const uint32_t offsets_from_utf8[6] = {
+  0x00000000, 0x00003080, 0x000E2080,
+  0x03C82080, 0xFA082080, 0x82082080
+};
+
+
+template <ErrorStrategy OnError, typename GlyphStrategy>
+inline size_t Utf8Consume(
+    const uint8_t* const input,
+    const size_t length,
+    const GlyphStrategy OnGlyph) {
+  size_t idx = 0;
+  while (idx < length) {
+    size_t advance = 0;
+    uint32_t glyph = 0;
+    uint8_t extrabytes = input[idx] ? clz(~input[idx]) : 0;
+    size_t i = idx;
+
+    if (extrabytes + idx > length) {
+      advance = OnError(length - idx, input, extrabytes);
+    } else if (!IsLegalUtf8Glyph(input + idx, extrabytes + 1)) {
+      advance = OnError(length - idx, input, extrabytes);
+    } else {
+      ASSERT(extrabytes < 4);
+      switch (extrabytes) {
+        case 3:
+          glyph += input[i++];
+          glyph <<= 6;
+          // fall-through
+        case 2:
+          glyph += input[i++];
+          glyph <<= 6;
+          // fall-through
+        case 1:
+          glyph += input[i++];
+          glyph <<= 6;
+          // fall-through
+        case 0:
+          glyph += input[i];
+      }
+
+      glyph -= offsets_from_utf8[extrabytes];
+
+      if (glyph > UNI_MAX_LEGAL_UTF32 ||
+          (glyph >= UNI_SUR_HIGH_START && glyph <= UNI_SUR_LOW_END)) {
+        advance = OnError(length - idx, input, extrabytes);
+      } else {
+        advance = extrabytes + 1;
+        OnGlyph(input + idx, extrabytes + 1, glyph, idx);
+      }
+    }
+
+    if (advance == 0) {
+      break;
+    }
+    idx += advance;
+  }
+  return idx;
+}
+
+
+size_t Utf8Value::StripInvalidUtf8Glyphs(uint8_t* const input, const size_t size) {
+  size_t idx = 0;
+  auto on_glyph = [input, &idx](
+      const uint8_t* data, size_t size, uint32_t glyph, size_t pos) {
+    size_t old_idx = idx;
+    idx += size;
+    if (old_idx == pos) return;
+    memmove(input + old_idx, data, size);
```

Comment:
It looks like repeatedly stripping bad character sequences can result in approx. O(n^2) bytes getting copied?